### ISSUE TITLE
Can't remove meters after a filter is added

### ISSUE THREAD

**Issue by echatman** [micrometer-metrics/micrometer#1597]
`MeterRegistry.remove()` fails to remove metrics that were created before a `commonTags()` filter was applied. It seems to apply the filter to the metric ID before trying to look it up, and so it finds nothing, and fails to delete it.

Tested using the latest version (`io.micrometer:micrometer-core:1.2.1`).

Here is my test case:
```java
import io.micrometer.core.instrument.Metrics;
import io.micrometer.core.instrument.Timer;
import org.junit.Test;

import java.time.Duration;

import static org.junit.Assert.assertNull;

public class RemoveMetricTest {

    /**
     * This passes as expected
     */
    @Test
    public void testRemoveMetric() {
        Timer testTimer = Metrics.globalRegistry.timer("testTimer");
        testTimer.record(Duration.ofMillis(10));
        Metrics.globalRegistry.remove(testTimer);
        assertNull(Metrics.globalRegistry.find("testTimer").timer());
    }

    /**
     * This fails - the Timer does not get removed
     */
    @Test
    public void testRemoveMetricAfterAddingFilter() {
        Timer testTimer = Metrics.globalRegistry.timer("testTimer");
        testTimer.record(Duration.ofMillis(10));
        Metrics.globalRegistry.config().commonTags("foo", "bar");
        Metrics.globalRegistry.remove(testTimer);
        assertNull(Metrics.globalRegistry.find("testTimer").timer());
    }
}
```

**Comment by izeye** [comment]
@echatman Thanks for the report! I'm not sure whether configuring meter filters dynamically was intended or not as the current implementation doesn't seem so.

**Comment by mweirauch** [comment]
(Sorry for closing the issue yesterday. Mobile was slower then my fingers).

I am unsure if this setup really should be an issue or not.

Given the case one would like to support common atgs before and after registrations, it would either mean having to search `Meter.Id` with and without common tags on removal which could lead to ambiguity and higher processing duration, or it would mean to remember for each registered `Meter` with which set of tags it was registered first.

I am pretty confident that both things are not desired from a "footprint" point of view. Additionally the most prominent Micrometer user - Spring Boot - makes sure to apply any `MeterFilter` before any `Meter` registration. The same modus should be followed when manually wiring the `MeterRegistry` setup.

**Comment by shakuzen** [comment]
I agree that things are wonky right now for dynamically adding filters. We should at the least make it more clear in documentation that filters are intended to be configured before registering any meters. As pointed out by others, it would not likely be easy to support this use case, and I'm not sure it is generally worth it without more background on why the use case is important.

Beyond finding a meter after calling remove in this case, your referenced `testTimer` is also probably not what you're expecting - it will not have the common tags configured on it. If you called `meterRegistry.timer("testTimer")` again after configuring the common tags, you would get a different `Timer` instance that has the common tags on it.

**Comment by cykl** [comment]
I just faced a similar issue with a slightly different setup. 

My application registers meters using the globalRegistry. Two MeterRegistry are added to the globalRegistry at startup, each registry has its own MeterFilter. When the application removes a meter,  it is removed from the CompositeRegistry but not from the two terminal registries has a meter have a different Meter.Id in each registry. I can open a dedicated issue if it helps.

**Comment by admoca60** [comment]
Dear all!! In my case, I wrote this comment at micrometer channel yesterday.

Dear people!
I'm using SB2.3.4 for my micro and I found a posible issue, but I'm not sure, because in case it was issue, I don't know exactly how to fix it. Following the description of the environment and the issue:
I have two different registries (in my example they are InfluxDB and Prometheus, but it can be reproducible for any pair of registries). Both registries are childs of CompositeMeterRegistry and I'm using this one (composite for add and remove metrics). Up to now, all is running well.
My problem raise when I configure a particular child registry to avoid some tags with the line
`influxMeterRegistry.config().meterFilter(MeterFilter.ignoreTags("tagkey2"));`
After this individual configuration of the child registry, the meters are correctly added, but there are not removed from InfluxMeterRegistry because the meterId are not exactly equals.
In my example, the metric is added to compositeMeterRegistry and prometheusRegistry with tags (tagkey and tagkey2) but in influxMeterRegistry, the metric is added with only "tagkey" (without "tagkey2", obvious because the meter filter avoid it).
In the removal process, the class "MeterRegistry" tries to delete from the meterMap taking into account the meterId which "equals" method implementation taking into account the name and the tags . It's for that, when try to remove the metric in influxMeterRegistry, the metric is not found and so it isn't deleted.
My question now: do you think that it can be considered as a bug? In my opinion, it not a bug strictly but, how can we remove child meters with different tags (autogenerated from a meter filter)? Maybe, the solution might be that any child meter should have a reference to parent metrics and in that case all child metrics will be removed when the parent meter is removed?
PS: I have the project to reproduce it. I'm going to upload to github if anyone need it  [Github Repository with use case](https://github.com/admoca60/childRegistriesMicrometerIssue)

**Comment by admoca60** [comment]
There is another issue related to this matter. The issue is #1159

**Comment by shakuzen** [comment]
> When the application removes a meter, it is removed from the CompositeRegistry but not from the two terminal registries as a meter has different Meter.Id in each registry. I can open a dedicated issue if it helps.

@cykl We have recently had #2354 opened which I believe matches the issue you described. Let's track that issue there.
@admoca60 the issue you described also sounds like the same one as that. It is different from the original scenario described here which only involves one registry.

**Comment by admoca60** [comment]
@shakuzen exactly, it sounds like #2354 and also #1159 which I have the PR (https://github.com/micrometer-metrics/micrometer/pull/2281 ) to solve this.

**Comment by admoca60** [comment]
Sorry @shakuzen , I'm aware just now that I failed in my analysis.

#2354 and this (#1597) are the same. They involves the meters.
#1159 involves the registries.

**Comment by admoca60** [comment]
> There is another issue related to this matter. The issue is #1159

It's not true. I was confused

**Comment by echatman** [comment]
I believe this is fixed along with #2354 - the test case in my original report now passes using version 1.6.3.

**Comment by admoca60** [comment]
> I believe this is fixed along with #2354 - the test case in my original report now passes using version 1.6.3.

Exactly, as I wrote, the #2354 and #1597 are the same use case!
